### DISCUSSION TITLE
Remove python/poetry venv autocreate feature

### DISCUSSION THREAD

**Discussion by jdx** [jdx/mise#1357] [Polls]
When I built the venv autoactivate I copied direnv's behavior which creates the venv or runs `poetry install` if it hasn't been done yet. I'm wondering if this is what we want though. Would it be simpler if rtx didn't do this and instead just activated an existing venv?

To be clear, I'm proposing removing the "creation" step, not the automatic activation (which I know is well liked). The creation is often the source of problems. Basically removing the `python -m venv` step and showing a warning if it doesn't exist.

Perhaps this could go behind a setting like `MISE_PYTHON_VENV_AUTO_CREATE` defaulting to disabled?

My thinking here is that because I'm not a python developer I often get the logic around venvs wrong. I'm simply not the best person to lead a project for python DX (just an unfortunate reality). I'm thinking rather than me make mistakes, just have mise do a bit less so I don't get the logic incorrect. I feel like probably 25% of mise issues are venv related. I often will fix it in one place but break it in another.

I'm leaning towards doing this over not doing this, maybe with some sort of tbd deprecation message. I very much welcome input from python users though.

On a related note, I also plan to move this config from the `[tools]` section to the `[env]` section like this:

```toml
[env]
mise.python.venv = ".venv"
mise.poetry.pyproject = "pyproject.toml"
```

My thinking this is more of an "environment" config and mostly unrelated to the tool (they are called "virtual environments" after all). This will also make it possible to use it with the system python which can't work in `[tools]` without significantly changing the meaning of the term "system" in mise.

Related issues:
* #1354 
* #1248 
* #1058 
* #1303

**Comment by jdx** [comment]
I'll be truly shocked if anyone picks the first option here. If you really think that's the solution you're going to need to back it up :)

**Comment by mustafa0x** [comment]
> On a related note, I also plan to move this config from the `[tools]` section to the `[env]` section like this:
> 
> ```toml
> [env]
> "@python:venv" = ".venv"
> ```

Isn't `[env]` for "setting arbitrary environment variables"? This usage seems to be different though.

**Reply by jdx** [reply]
if I remove the autocreation, then the only thing that would do is effectively set `VIRTUAL_ENV=.venv` with a bit of logic to show a warning if it doesn't exist.

"@" and ":" are invalid characters in env var keys[^keys] so I plan to use that syntax to denote special behavior within env var config like this.

I would argue even with autocreation putting it in `[tools]` is the wrong place. It's not managing the python runtime at all.

[^keys]: I think technically you _can_ potentially use these characters in env vars at least on some OS's, but it's not well supported in any shell I don't think so it's practically not supported.

**Reply by jdx** [reply]
update: changed the toml syntax a bit to make it not require quotes

**Reply by mustafa0x** [reply]
Ok, so it'll just be a shorthand for setting a special environment variable. Makes sense, and that doesn't feel too much like implicit behavior / magic / overloading. Thanks.

**Reply by delicb** [reply]
I can stand behind this format. But `VIRTUAL_ENV` is not enough, `PATH` should be updated as well.

**Reply by jdx** [reply]
> I can stand behind this format. But VIRTUAL_ENV is not enough, PATH should be updated as well.

yeah exactly right. There is very good support in the internals of mise for modifying PATH though so it's easy to make that work

**Comment by delicb** [comment]
Adding a setting with default being disabled would be fine (option no. 3) IMO. This is the option I am voting for. 

However, I just had a different thought. I did not follow development for some time (just used it as is, until rename brought me back to this repo) so I did not know that there is not experimental support for tasks. As far as I understand, tasks are not executed directly. However, if there is a way to:
* execute a task on event (e.g. cd into directory and cd out)
* export environment variables from task

Then special handling of python virtual environments could boil down to something like:
```toml
[tasks.pyvenv]
description = 'Create and activate virtual env'

# you can specify a multiline script instead of individual commands
run = """
#!/usr/bin/env bash
[ ! -d venv ] && python -m venv .venv
source .venv/bin/activate
"""
export_vars = ["VIRTUAL_ENV", "PATH"]
run_on = ["enter"]
```

Of course, example above does not work now (`export_vars` and `run_on`), but if such features were possible, this could be used to tailor behavior as needed. Having ability to "call" some globally defined task (so that same task does not have to be repeated in every `.mise.toml` (this is inspired by `direnv`, which I was using heavily until I switched to this tool).

Again, I have no idea what the roadmap for tasks feature is, but I have a filling it could be helpful here and leave only bare python version management for `mise` core.

**Reply by jdx** [reply]
what you're describing is a feature we don't have, I think I would call it "hooks" if we did implement it. Tasks are a different use-case, they're for manually executed things where you explicitly run `mise run ...`. Still, that's semantics, I could see us adding something exactly like this but we don't have it today.

There was just some conversation around a similar idea to build env vars based on external commands too which might be relevant: https://github.com/jdx/mise/discussions/1301

Templates even enable this today with some caveats (performance can be impacted since they get run basically every time mise starts)

**Reply by delicb** [reply]
Yep, I understand that there is no such a feature today, I was just imagining how it would look like and if it could be helpful in this discussion. "Hooks" is a good name for it. I assume that a lot of implementation with tasks would be shared. 

Event if it is not there today, I imagine this is such a pressing issue that it has to be solved right now, it could wait for "hooks", if they are on roadmap.

**Comment by jdx** [comment]
wow unanimous so far. So I'll go with disabling by default unless some config is set. I think with this in place I think I'm finally ready to label it as non-experimental anymore.

Follow-up question though: what setting do you plan to use?

**Reply by AlexCzar** [reply]
In my use-cases it'll be mostly on by default. `.mise.toml` files are committed to repos and when I clone the repo it's nice that I don't have to do any manual setup, I just cd into it and mise takes care of everything.

I also broke the unanimity, sorry :). My argument is not my own use-case though, but backward-compatibility.
The behaviour is currently to auto-create, adding the setting arguably should break as little things as possible, which means that new setting's default should also be to auto-create.

Additionally, I want to mention a couple of issues I'm experiencing with the currently existing functionality.
It would be nice if they are addressed as part of the planned refactoring.

1. I have a project which has .mise.toml at the root level, which specifies python version and venv.
I have a subdirectory in that project that has nothing to do with python but addresses some other aspect of the project (say, deployment). When I add `.mise.toml` to the subpdirectory, mise wants to set up venv there, too, even though that subdirectory's `.mise.toml` doesn't have any python stuff at all.
2. Mise tasks ignore venv. I tried defining something like `python -m code_formatter` as a mise task where code_formatter module is installed in the venv, but this fails because the python that runs is the base one, not the venv one. This is easy enough to work around, but annoying. I assume that everything that mise configures should be accessible to its tasks, like environment, PATH, etc.

**Reply by jdx** [reply]
> I also broke the unanimity, sorry :). My argument is not my own use-case though, but backward-compatibility.

keep in mind this is an experimental feature explicitly to be clear that breaking changes should be expected

**Reply by jdx** [reply]
I am pretty sure moving the config to `[env]` will resolve your first issue. It will mean the tool config and virtualenv config can be in separate files.

I'm actually a little surprised your second one is happening—it definitely should be using the venv one. I'll need to investigate that but it just sounds like a straightforward bug. I'm probably not using the right variable or initializing the right thing.

**Reply by AlexCzar** [reply]
1. Totally, I just mentioned that breakage as an argument as to why I feel keeping auto-create on by default would be nice.
2. Good to hear that :)

**Comment by noirbizarre** [comment]
My 2 cts: I think moving this to `env` is a bad idea: initially `env` was for "environment variables": clear scope, well known naming.
Now it accepts some specials `mise.*` vars for env_files. While this is great to be able to support multiple env vars, it makes the scope and purpose of `env` blurry.

Adding `virtualenv` configuration into it while it's not strictly related to environment variables will just open the door to having a one-catch-all `env` section and add lots of confusion.

By consistency with the environment variable `MISE_TOOL_OPTS__VIRTUALENV` and thinking forward compatibility, I think `virtualenv` management is some tooling/plugin config and should go in a dedicated section for this tool, something like:

```toml
[tools]
python = "3.7"

[python]
virtualenv = ".venv"
```

I think it's more scalable for both other Python specific options or other plugin that will for sure have some specific fine-tuning to add later.

It also respects separation of concerns.

Note that while part of virtualenv management is about hijacking some environment variables, it's not just that. How can I specify that I want to have system-packages in the venv ? Another special syntax ?
What if we want to support some pre-installed dev tools ? What if later we want to distinguish "reuse virtualenv" and "create if missing" ?

The dedicated section allows this:
```
[python]
virtualenv = ".venv"
create-virtualenv = true
site-packages = true
packages = ["ipython", "debugpy"]
```
This would be harder to handle with special env processing.

In the case of Python, adding some `poetry`, `pdm` or any package manager specific settings would be easier as both same level options or nested table.

At last having a dedicated section with explicit options make it possible to have a clear schema for completion or validation.

And I didn't raise the fact that it introduce an order and precedence issue: when does the venv get activated: before env, after env processing, at the same time that other variables... ?


*Note:* having the possibility to have `env` and `env_files` processed before and after plugins activations would be just perfect and allow:

```toml
[env]
SOME_VAR_THAT_MIGHT_BE_NEEDED_BEFORE_ACTIVATION = "value"

[tool]
python = "3.11"

[python]
virtualenv = ".venv"

[lazyenv]  # just for the example, naming TBD
SOME_VAR = "{{ env.VIRTUAL_ENV }}/some/extra/path"
```

**Reply by jdx** [reply]
thanks for the details response @noirbizarre

From a high level: my goal is to separate concerns with mise along 3 axes: tools, environment, and tasks. The docs and the config will be organized around these 3 concepts. What I'm trying to avoid is top-level keys like `[python]` where it's not clear to the user which of those 3 things it falls under. It's hard for me to overstate just how important I see this docs/config alignment.

I want homogeneity within these 3 sections. That's trivial with tasks. With tools is slightly more complicated but still not too bad. With env, I think I'm on the right track here, but it's admittedly not something I think people see at first glance.

I think if you just see mise as simply a way to manage your venv and set some env vars as separate things then yes, I agree with you that having a dedicated `[python]` section is more straightforward—but that's not my goal here. My goal is to _also_ teach people how mise works under the hood so when they want to branch out to something like running a custom script to generate their own env vars it'll look familiar because it's similar to how they setup a venv or added a dotenv file.

It's a slight hit to DX up front, but it helps users along the path to building more complex config files in the direction I want them to head.

The way I see `[env]` are individual modules that set environment variables potentially with side effects. For example, you should be able call a script to generate env vars:

```toml
[env]
mise.script = "./get_my_env_vars.sh"
```

Or use a plugin:

```toml
[env]
mise.nix.flake = "flake.nix"
```

"mise." here is a directive that signifies what you're defining is not a simple KV but a separate module. Perhaps that syntax could be changed, I had considered going with this instead:

```toml
[env]
"@nix".flake = "flake.nix"
```

But I don't like that it needs to be quoted like that. `mise.` was the best thing I could think of that won't require quotes.

> My 2 cts: I think moving this to env is a bad idea: initially env was for "environment variables": clear scope, well known naming.
Now it accepts some specials mise.* vars for env_files. While this is great to be able to support multiple env vars, it makes the scope and purpose of env blurry.

Part of the issue is using `[python]` doesn't give me any way to have "builtin" modules like `mise.file`, `mise.script`, and `mise.path` without those all being their own top-level things.

> And I didn't raise the fact that it introduce an order and precedence issue: when does the venv get activated: before env, after env processing, at the same time that other variables... ?

That's the argument for putting this into `[env]` and _not_ giving plugins their own section. You should be able to do this:

```toml
mise.python.virtualenv = '.venv'
MY_VENV = '{{env.VIRTUAL_ENV}}'
```

While possible to still do this with multiple `[env]` sections, I think that's a lot uglier and has some parsing problems.

> Another special syntax ?

I think the syntax would be similar for both scenarios. Tons of places in config parsing already accept either a simple value like a string or a map of options:

```toml
[env]
mise.python.virtualenv = {path=".venv", default-packages=[], create=true}

[python]
virtualenv = {path=".venv", default-packages=[], create=true}
```

> At last having a dedicated section with explicit options make it possible to have a clear schema for completion or validation.

Schemas are unaffected by this decision. `env.mise.file|path` are already validated.

**Reply by jdx** [reply]
another thing that is less important to you all, but I think makes me think I'm headed in the right direction is that this will make the codebase more streamlined. What I would do is implicitly load the tools as the first step in `[env]`—that way "tools loading" is just another `[env]` module. The user won't see that, but it makes the behavior a lot more obvious.

**Reply by jdx** [reply]
Maybe if I tried to summarize what I'm saying here: I see `[env]` as a "platform". It's for me building the core, plugin authors, and users to build their own "modules" inside of. They could mix and match them, change the order, create custom modules, etc. I know people will do this because they already are—doing crazy things with templates that execute scripts. I want to make that pattern consistent across everything in mise that sets environment variables so it's clear exactly how env vars work in mise.

**Reply by noirbizarre** [reply]
> Part of the issue is using [python] doesn't give me any way to have "builtin" modules like mise.file, mise.script, and mise.path without those all being their own top-level things.

I might be missing something because I don't see how it's related to python or virtualenv activation. And more, I don't see how the problem you raise disappear in `env`:
- either it is at the `env` level, and it will only work for `env`
- either it works for any subkey of `env` like `env.python.mise.script` and you just have the exact same problem than with a root key which would give `python.mise.script` if you want to reuse those.

This is `toml` so literally those 3 syntaxes are strictly equivalents:

```toml
[env]
mise.python.virtualenv = {path=".venv", default-packages=[], create=true}
```

```toml
[env.mise.python]
virtualenv = {path=".venv", default-packages=[], create=true}
```

```toml
[env.mise.python.virtualenv]
path = ".venv"
default-packages = []
create = true
```

And to me, none of those is as simple and explicit than

```toml
[python]
virtualenv = ".venv"
virtualenv-create = true
default-packages = []
```

I know this is personal taste, but still, I don't get the point of having a `env.mise.python` subsection in a `.mise.toml` file to declare something that would have been strictly equivalent with clearer keys at root level.

> From a high level: my goal is to separate concerns with mise along 3 axes: tools, environment, and tasks. The docs and the config will be organized around these 3 concepts. What I'm trying to avoid is top-level keys like [python] where it's not clear to the user which of those 3 things it falls under. It's hard for me to overstate just how important I see this docs/config alignment.

Then why not keep `tools.{tool}` it is already consistent with your 3 axes visions. This way, we can have both simple and advanced forms:

```toml
[tools]
python = "3.11"
```
or
```toml
[tools.python]
version = "3.11"
virtualenv = ".venv"
```

But please 🙏🏼 , don't make tooling conf in `env`. For users with lots of env vars it will just make a big unreadable section where you will have to do visual triage to get your info while separated sections are just easier to read, easy to query, easy to remember. All in env, we lose the visual management unless we start putting comments everywhere.


Note that I chose to answer because in a few weeks we are moving from "just the perfect tool having just what I need, basic simple efficient and readable" to renaming+refactoring in a direction that seems different. I am not here to throw stones or something else, but because I fear to trade readability and efficiency for complexity. Because, I think `mise` config file is good because it is straight forward and easy to learn/underst.

Anyway, I just wanted to share hoping there might be a chance that you change your mind 🙏🏼

(I love `mise` and the way you manage and interact with its community and don't want to offend anybody, just trying to voice my case. In fine your soft and your vision, I'm OK with that and I respect that. Just so there is no ambiguity)

**Reply by jdx** [reply]
So I wouldn't fret too much. env_path/env_file I think were unchanged for almost a year. Those were high on my list of things to change and I spent a lot of time thinking of a new design for them. That said, virtualenv activation is marked as experimental. Don't use that (or any experimental feature) if you don't want to deal with breaking changes. Me marking something experimental is me telling you that I plan to modify the design and I am not going to bother with a deprecation path.

I don't think anything I'm proposing here would result in breaking changes to non-experimental features. If it is, then that's a problem. Also, my plans around environments here I really don't know if I'll get to for maybe a year. I want to get an idea in my head about what I want the end-state to look like which is what I'm doing in these discussions.

Most configuration in mise I see as set in stone. I don't think I would ever change how `[env]` works for static variables. Since it doesn't make sense to nest env vars, `[env.mise.*]` won't break anything existing. Tools will get extended but none of the current (non-experimental) features will change. `[tasks]` is going to change a lot but that's very new and experimental.

When it comes to venv though, there are things I know and things I don't know. What I do know is `[tools]` is the wrong place for venv. I'm also fairly convinced that a top-level `[python]` is wrong. `[tools.python]` makes it so you can't put venv config into .mise.local.toml without duplicating the runtime config. It doesn't work with system installations. It's also just "wrong" from a purity standpoint—it's attached to the runtime but it really doesn't have anything to do with it. So I think `[tools.python]` is a non-starter.

I have considered a top-level `[python]` a few times but I think I'm unlikely to ever do that because to me that would imply that the runtime configuration should also be there, e.g.:

```toml
[python]
version = "3.11"
virtualenv = ".venv"
```

Potentially in an alternate universe mise could've been designed that way—language first—but it wasn't. Given that, I don't see this as an option. `[python]` and `[tools.python]` does not mix.

`[env.mise.python]` or `[env.python]` I think is the right place for that logic. You're not at all wrong that this does make the model a little weird with strings at the top level and maps lower down. Maybe a new top-level key would be better like `[env_loaders]`. I've been trying to think of a way that might work but so far I haven't thought of a key that works.

Another problem with `[env.mise.python]` isn't a problem for venvs which only would ever have a single one, but for something like `[env.mise.script]` it's not clear to me how someone might use multiple scripts.

`[[env]]` is a band-aid solution but I think it's ugly.

Anyhow, I'm still open to ideas, but `[tools.python]` I've certainly disregarded and `[python]` I'm very much leaning against. I don't think that `[env.mise.python]` is perfect but I do think it's feasible I don't think the critiques are dealbreakers.

**Reply by noirbizarre** [reply]
OK, I reread it all, I missed the order and precedence point. Now I get it and it makes sense 👍🏼.

But I realize it means that given they are toml keys `mise.script`, `mise.path` and `mise.file` can only be loaded/executed once by design. (It's already the case but `mise.script` feels like something that could be executed multiple times with different scripts)

**Reply by jdx** [reply]
Also thanks again for the discussion. I'm going to probably read your posts over again a few times over the coming days. If anything is clear, I'm not opposed to changing my mind—arguably to a fault :)

**Reply by jdx** [reply]
> But I realize it means that given they are toml keys mise.script, mise.path and mise.file can only be loaded/executed once by design. (It's already the case but mise.script feels like something that could be executed multiple times with different scripts)

Yeah you've hit on the biggest problem with `[env.script]`. I think there is potentially room to use arrays there, maybe `[[env]]` or `[[env.script]]`. I haven't fleshed that thought out yet though.

**Reply by noirbizarre** [reply]
Thanks for taking time to respond (and read!) 🙏🏼 That's an interesting discussion for sure.

Now that I have a better understanding, here's a suggestion for the multiple scripts case (that just cross my mind):
- given I don't think it will be the most common case, I'd keep `env.mise.script`
- and for those wanting multiscripts capabilities, a `env.mise.scripts` table, like named scripts where name is just ignored (a tradeoff let's say)

Also, to reduce verbosity and provide a visual distinction, why not use `env._` ?

```toml
[env]
_.scripts.first = "script1.sh"
_.python.virtualenv = ".venv"
_.scripts.second = "script2.sh"
MY_VAR = "{{ env.VIRTUAL_ENV }}/somewhere"

[tools]
python = "3.11"
```

**Reply by jdx** [reply]
I do prefer `_` over `mise`. I think it's clear it's a directive. I think this could work. I did have a few other thoughts (that I don't necessarily prefer):

```toml
[[env._.script]]
run = "script1.sh"
[env]
_.python.virtualenv = ".venv"
MY_VAR = "{{ env.VIRTUAL_ENV }}/somewhere"
[[env._.script]]
run = "script1.sh"
[tools]
python = "3.11"
```

what's nice about this is you don't need to make up a name like "first/second". I don't love that you can't do it inside of the `[env]` block though. I also think `[[env._.script]]` looks really weird and probably nonsense to someone that doesn't really understand toml.

```toml
[env]
_.scripts = "script1.sh"
_.python.virtualenv = ".venv"
_.scripts_ = "script2.sh"
MY_VAR = "{{ env.VIRTUAL_ENV }}/somewhere"

[tools]
python = "3.11"
```

this is easy for me to parse since I would just strip off any extra "_" at the end of keys.

**Reply by jdx** [reply]
or I was also thinking of just making `[env]` itself optionally an array:

```toml
[[env]]
_.scripts = "script1.sh"
_.python.virtualenv = ".venv"

[[env]]
_.scripts = "script2.sh"
MY_VAR = "{{ env.VIRTUAL_ENV }}/somewhere"

[tools]
python = "3.11"
```

**Reply by jdx** [reply]
I kind of think this last one is the cleanest, any thoughts?

**Reply by jdx** [reply]
I think I might change the deprecation method for both the existing env_path and if we change from env.mise to env._.

There isn't any good reason to warn the user. I'll change the docs but keep parsing the old syntax indefinitely (or at least for a very long time).

That's normally how I deprecate things. I think with mise showing an actual warning should be reserved for things actually likely to cause a problem. Not just something to simply improve consistency.


Consistency is important, but I think in this case the remedy is worse.

**Reply by noirbizarre** [reply]
> I do prefer `_` over mise. I think it's clear it's a directive. I think this could work.

Glad I shared the idea !

> or I was also thinking of just making `[env]` itself optionally an array:
> 
> ```toml
> [[env]]
> _.scripts = "script1.sh"
> _.python.virtualenv = ".venv"
> 
> [[env]]
> _.scripts = "script2.sh"
> MY_VAR = "{{ env.VIRTUAL_ENV }}/somewhere"
> ```

I do like the idea if it's optional so people not requiring multiple script can still use the basic form. Given it is totally possible to validate the file format, I don't have an issue with the 2 forms approach.

> ```toml
> [[env._.script]]
> run = "script1.sh"
> [env]
> _.python.virtualenv = ".venv"
> MY_VAR = "{{ env.VIRTUAL_ENV }}/somewhere"
> [[env._.script]]
> run = "script1.sh"
> [tools]
> python = "3.11"
> ```
> 

This one feels a bit to complicated to me. If we go the array way, `[[env]]` seems more natural and simple.

> what's nice about this is you don't need to make up a name like "first/second". I don't love that you can't do it inside of the `[env]` block though. I also think `[[env._.script]]` looks really weird and probably nonsense to someone that doesn't really understand toml.
> 
> ```toml
> [env]
> _.scripts = "script1.sh"
> _.python.virtualenv = ".venv"
> _.scripts_ = "script2.sh"
> MY_VAR = "{{ env.VIRTUAL_ENV }}/somewhere"
> 
> [tools]
> python = "3.11"
> ```
> 
> this is easy for me to parse since I would just strip off any extra "_" at the end of keys.

I do like this idea too and I though about it. It was the alternative to naming. `_.script*` are executed in order or appearance. It handles multiple scripts quite well and I don't believe we should aim for a solution that fit lots of scripts. If people need lots of scripts, chances are it can be merged as a single or 2. Multiple script are just there to be able to have scripts/module loading alternance, not to replace a shell so most of the time people will just have `_.script`, `_.script_` and maybe `_.script_X` where X is a number. Good enough to handle has many cases as it should I think.
Note that I used the singular form. If we go the `_.script*` prefix, we don't need to have a separate plural form and for the basic 1 script case, `_.script` seems more fitted. But it's a detail, plural form works too.

**Reply by jdx** [reply]
right now I think I'm leaning towards `[[env]]` as the solution—of course `[env]` would continue to work as it does today. I would probably just put this in the docs under a "tip" callout or something that you can use `[[env]]` instead in the situation where you have multiple scripts.

Thankfully we've got time to come to a decision here, as you said this is an edge case and `_.scripts` isn't even supported yet.
